server/src/main/java/org/apache/iotdb/db/auth/AuthorizerManager.java

```diff
@@ -215,131 +245,312 @@ public void reset() throws AuthException {
     try {
       return iAuthorizer.getAllUserWaterMarkStatus();
     } finally {
-      snapshotLock.readLock().unlock();
+      authReadWriteLock.readLock().unlock();
     }
   }
 
   @Override
   public Map<String, User> getAllUsers() {
-    snapshotLock.readLock().lock();
+    authReadWriteLock.readLock().lock();
     try {
       return iAuthorizer.getAllUsers();
     } finally {
-      snapshotLock.readLock().unlock();
+      authReadWriteLock.readLock().unlock();
     }
   }
 
   @Override
   public Map<String, Role> getAllRoles() {
-    snapshotLock.readLock().lock();
+    authReadWriteLock.readLock().lock();
     try {
       return iAuthorizer.getAllRoles();
     } finally {
-      snapshotLock.readLock().unlock();
+      authReadWriteLock.readLock().unlock();
     }
   }
 
   @Override
   public void replaceAllUsers(Map<String, User> users) throws AuthException {
-    snapshotLock.readLock().lock();
+    authReadWriteLock.readLock().lock();
     try {
       iAuthorizer.replaceAllUsers(users);
     } finally {
-      snapshotLock.readLock().unlock();
+      authReadWriteLock.readLock().unlock();
     }
   }
 
   @Override
   public void replaceAllRoles(Map<String, Role> roles) throws AuthException {
-    snapshotLock.readLock().lock();
+    authReadWriteLock.readLock().lock();
     try {
       iAuthorizer.replaceAllRoles(roles);
     } finally {
-      snapshotLock.readLock().unlock();
+      authReadWriteLock.readLock().unlock();
     }
   }
 
   @Override
   public boolean processTakeSnapshot(File snapshotDir) throws TException, IOException {
-    snapshotLock.writeLock().lock();
+    authReadWriteLock.writeLock().lock();
     try {
       return iAuthorizer.processTakeSnapshot(snapshotDir);
     } finally {
-      snapshotLock.writeLock().unlock();
+      authReadWriteLock.writeLock().unlock();
     }
   }
 
   @Override
   public void processLoadSnapshot(File snapshotDir) throws TException, IOException {
-    snapshotLock.writeLock().lock();
+    authReadWriteLock.writeLock().lock();
     try {
       iAuthorizer.processLoadSnapshot(snapshotDir);
     } finally {
-      snapshotLock.writeLock().unlock();
+      authReadWriteLock.writeLock().unlock();
+    }
+  }
+
+  public TSStatus checkPath(String username, List<String> allPath, int permission)
+      throws AuthException {
+    authReadWriteLock.readLock().lock();
+    try {
+      User user = userCache.getIfPresent(username);
+      boolean status = true;
+      if (user != null) {
+        for (String path : allPath) {
+          if (!user.checkPrivilege(path, permission)) {
+            for (String roleName : user.getRoleList()) {
+              Role role = roleCache.getIfPresent(roleName);
+              // It is detected that the role of the user does not exist in the cache, indicating
+              // that the permission information of the role has changed.
+              // The user cache needs to be initialized
+              if (role == null) {
```

Comment:
Although the role of the user is changed, we should still judge the permission using new role ? not sure...

Reply:
Got it, the retry logic is in line 412, is it correct ?

Reply:
yes